### PR TITLE
SYNCT-189: Added whitelist validation & conversions for wrapped messages

### PR DESCRIPTION
**Pull request by wojtekbu**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/sync2/api/service/impl/SyncRequestWrapperServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        return isInstanceIdValid(requestWrapper.getInstanceId());
+    }
+
+    private boolean isInstanceIdValid(String instanceId) {
+        return isWhitelistDisabled() || isInstanceIdOnWhitelist(instanceId);
+    }
+
+    private boolean isInstanceIdOnWhitelist(String instanceId) {
+        return configuration.getSyncConfiguration().getWhitelist().getInstanceIds().contains(instanceId) &&
+                configuration.getSyncConfiguration().getWhitelist().isEnabled();
```
Comment (oskarhinc) [inline api/src/main/java/org/openmrs/module/sync2/api/service/impl/SyncRequestWrapperServiceImpl.java]: configuration.getSyncConfiguration().getWhitelist().isEnabled() could be !isInstanceDisabled()

File: api/src/test/java/org/openmrs/module/sync2/api/sync/SyncClientTest.java [hunk 2]
```diff
@@ -60,13 +59,18 @@
     private static final String PATIENT_CATEGORY = "patient";
     private static final String VISIT_CATEGORY = "visit";
     private static final String PARENT_ADDRESS = "http://localhost:8080/";
+    private static final String CHILD_ADDRESS = "http://localhost:8080/";
```
Comment (oskarhinc) [inline api/src/test/java/org/openmrs/module/sync2/api/sync/SyncClientTest.java]: are parent and child addresses equal?

File: api/src/main/java/org/openmrs/module/sync2/client/rest/RESTClientHelper.java [hunk 3]
```diff
@@ -97,6 +108,31 @@ public boolean compareResourceObjects(String category, Object from, Object dest)
...
+    }
+
+    @Override
+    public String convertToFormattedData(Object object) {
+        if (RestResource.class.isAssignableFrom(object.getClass())) {
+            return restConverter.convertToJson((RestResource) object);
+        } else if (SimpleObject.class.isAssignableFrom(object.getClass())) {
+            return simpleConverter.convertToJson((SimpleObject) object);
+        } else {
+            throw new UnsupportedClassVersionError(String.format("Class %s is not supported",
```
[the remark this answers is not in the document]
Comment (wojtekbu) [reply api/src/main/java/org/openmrs/module/sync2/client/rest/RESTClientHelper.java]: OK

File: api/src/test/java/org/openmrs/module/sync2/api/sync/SyncClientTest.java (hunk 2 above)
Comment (wojtekbu) [reply api/src/test/java/org/openmrs/module/sync2/api/sync/SyncClientTest.java]: Unnecessary var

File: api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java [hunk 4]
```diff
@@ -41,19 +47,27 @@
 
 	private String password;
 
-	private RestTemplate restTemplate = new RestTemplate();
+	private RestTemplate restTemplate;
+
+	public SyncClient() {
+        restTemplate = new RestTemplate();
```
[the remark this answers is not in the document]
Comment (wojtekbu) [reply api/src/main/java/org/openmrs/module/sync2/api/sync/SyncClient.java]: I'll join it with the declaration

File: api/src/main/java/org/openmrs/module/sync2/client/rest/RESTClientHelper.java [hunk 5]
```diff
@@ -26,9 +30,14 @@
...
 	public static final String VOIDED = "voided";
 
-	private static final String ACCEPT_HEADER = "Accept";
+	private final RestHttpMessageConverter restConverter;
 
-	private static final String ACCEPT_MIME_TYPE = "application/json";
+	private final SimpleObjectMessageConverter simpleConverter;
+
+	public RESTClientHelper() {
+        restConverter = new RestHttpMessageConverter();
```
[the remark this answers is not in the document]
Comment (wojtekbu) [reply api/src/main/java/org/openmrs/module/sync2/client/rest/RESTClientHelper.java]: I'll join it with the declaration

**Pull request merged**